File: src/main/Java/dfs/boj11724.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.stream.IntStream;

public class boj11724 {
    private static boolean[] isVisited;
    private static ArrayList<Integer>[] linked;
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        int n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());
        isVisited = new boolean[n + 1];
        linked = new ArrayList[n + 1];
        IntStream.range(1, n + 1).forEach(index -> linked[index] = new ArrayList<>());
        for (int i = 0; i < m; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int x = Integer.parseInt(tokenizer.nextToken());
            int y = Integer.parseInt(tokenizer.nextToken());
            linked[x].add(y);
            linked[y].add(x);
        }
        int count = 0;
        for (int i = 1; i <= n; i++) {
            if (!isVisited[i]) {
                count++;
                dfs(i);
            }
        }
        writer.append(String.valueOf(count)).close();
    }
    private static void dfs (int n) {
        if (isVisited[n]) return;
        isVisited[n] = true;
        for (int i: linked[n]) if (!isVisited[i]) dfs(i);
    }
}

File: src/main/Kotlin/others/boj11724.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.stream.IntStream

private var linked = arrayOf(ArrayList<Int>())
private var isVisited = booleanArrayOf()
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val (n, m) = r.readLine().split(" ").map { it.toInt() }
    var count = 0
    linked = Array(n + 1) { ArrayList() }
    IntStream.range(0, n).forEach { linked[it] = ArrayList() }
    isVisited = BooleanArray(n + 1)
    repeat(m) {
        val (x, y) = r.readLine().split(" ").map { it.toInt() }
        linked[x].add(y)
        linked[y].add(x)
    }
    repeat(n) {
        if (!isVisited[it + 1]) {
            count++
            dfs(it + 1)
        }
    }
    w.append(count.toString()).close()
}

private fun dfs(n: Int) {
    if (isVisited[n]) return
    isVisited[n] = true
    linked[n].forEach { if (!isVisited[it]) dfs(it) }
}
